pkg/download/yamlcreator/yaml_creator.go

```diff
@@ -58,19 +88,176 @@ func (yc *YamlConfig) AddConfig(name string, rawName string) {
+				yc.Config[i][k] = jsonFileName
+			}
+		}
+	}
+
+	if !isExistingConfig {
+		config := map[string]string{}
+		config[configId] = jsonFileName
+		yc.Config = append(yc.Config, config)
+	}
+}
+
+func (yc *YamlConfig) getConfigName(entityId string) string {
+	for _, v := range yc.Detail {
+		config := v[0]
+		if config.Id == entityId {
+			return config.Name
+		}
+	}
+
+	return entityId
+}
+
+func (yc *YamlConfig) cleanseYamlConfig() CleansedYamlConfig {
+	yamlConfig := CleansedYamlConfig{
+		Config: []map[string]string{},
+		Detail: map[string][]CleansedDetailConfig{},
+	}
+
+	for k, v := range yc.Detail {
+		// isDownloaded specifies whether config was downloaded, i.e. currently exists in environment
+		// If it wasn't downloaded, it also won't be added to the config yaml.
+		isDownloaded := v[0].IsDownloaded
+
+		if isDownloaded {
+			entityName := v[0].Name
+			configFileName := v[0].ConfigFileName
+
+			yamlConfig.Detail[k] = []CleansedDetailConfig{}
+
+			cleansedDetailConfig := CleansedDetailConfig{
+				Name: entityName,
+			}
+
+			yamlConfig.Detail[k] = append(yamlConfig.Detail[k], cleansedDetailConfig)
+
+			cleansedConfig := map[string]string{}
+			cleansedConfig[k] = configFileName
+
+			yamlConfig.Config = append(yamlConfig.Config, cleansedConfig)
+		}
+	}
+
+	return yamlConfig
+}
+
 //CreateYamlFile transforms the struct into a physical file on disk
 func (yc *YamlConfig) CreateYamlFile(fs afero.Fs, path string, name string) error {
+	yamlConfig := yc.cleanseYamlConfig()
+	fullPath := filepath.Join(path, name+".yaml")
 
-	data, err := yaml.Marshal(yc)
+	data, err := yc.marshalYaml(yamlConfig)
 	if err != nil {
 		util.Log.Error("error parsing yaml file: %v", err)
 		return err
 	}
-	fullPath := filepath.Join(path, name+".yaml")
-	err = afero.WriteFile(fs, fullPath, data, 0664)
+
+	err = yc.writeFile(fs, fullPath, data, 0664)
 	if err != nil {
 		util.Log.Error("error creating yaml file %s", name)
 		return err
 	}
+
+	return nil
+}
+
+func (yc *YamlConfig) parseConfigs(unmarshaledData map[string]map[string]string) {
+	for k, v := range unmarshaledData {
+		if k == "config" {
+			for configId, configJson := range v {
+				configItem := make(map[string]string)
+				configItem[configId] = configJson
+
+				yc.Config = append(yc.Config, configItem)
+			}
+		}
+	}
+}
+
+func (yc *YamlConfig) findConfigFileName(configId string) string {
+	for _, v := range yc.Config {
+		for potentialId, configFileName := range v {
+			if potentialId == configId {
+				return configFileName
+			}
+		}
+	}
+
+	return ""
+}
+
+func (yc *YamlConfig) parseConfigDetails(unmarshaledData map[string]map[string]string) error {
+	for configId := range unmarshaledData {
+		if configId != "config" {
+			// As of May 2022, download does not support project structure
+			// Fallback to environment unique id
+			environmentUniqueConfigId := yc.EnvironmentName
+
+			entityUuid, err := yc.generateUuidFromConfigId(environmentUniqueConfigId, configId)
```

Comment:
In the case that the previously downloaded config is one with a unique name, the generated UUID will never match anything that Dynatrace actually returns. I don't see a need to generate any UUIDs here.

If I'm not missing something in the rest of the yaml_creator logic, I think you could replace this with: 
`if IsUuid(configId) then {use it as DetailConfig.Id} else {leave DetailConfig.Id blank}`

Reply:
If configId is no UUID, a new UUID is generated based on configId (unique within yaml file) + environment + project. This generated UUID is then used to create the config entity.